harness/ts/harness.ts

```diff
@@ -0,0 +1,138 @@
+import { Connection, WorkflowClient, WorkflowHandleWithRunId } from '@temporalio/client';
+import { ActivityInterface, Workflow } from '@temporalio/common';
+import { Worker } from '@temporalio/worker';
+import { promises as fs } from 'fs';
+import * as path from 'path';
+import { v4 as uuidv4 } from 'uuid';
+
+export interface FeatureOptions<W extends Workflow, A extends ActivityInterface> {
+  // Defaults to import(workflowsPath).workflow
+  workflow?: W;
+  // Defaults to no activities
+  activities?: A;
+  workflowsPath?: string;
+  execute?: (runner: Runner<W, A>) => Promise<WorkflowHandleWithRunId<W>>;
+  checkResult?: (runner: Runner<W, A>, handle: WorkflowHandleWithRunId<W>) => Promise<void>;
+  checkHistory?: (runner: Runner<W, A>, handle: WorkflowHandleWithRunId<W>) => Promise<void>;
+}
+
+export class Feature<W extends Workflow, A extends ActivityInterface> {
+  constructor(readonly options: FeatureOptions<W, A>) {}
+
+  public get activities(): A | ActivityInterface {
+    return this.options.activities ?? {};
+  }
+}
+
+export class FeatureSource {
+  static async findAll(absDir: string, origRootDir: string = absDir): Promise<FeatureSource[]> {
+    const dirents = await fs.readdir(absDir, { withFileTypes: true });
+    const dirs = [];
+    for (const dirent of dirents) {
+      if (dirent.name === 'feature.ts') {
+        const relDir = path.relative(origRootDir, absDir).replaceAll(path.sep, '/');
+        dirs.push(new FeatureSource(relDir, absDir));
+      } else if (dirent.isDirectory()) {
+        dirs.push(...(await FeatureSource.findAll(path.join(absDir, dirent.name), origRootDir)));
+      }
+    }
+    return dirs;
+  }
+
+  constructor(
+    // Relative to features/ root _and_ uses / for platform independence
+    readonly relDir: string,
+    readonly absDir: string
+  ) {}
+
+  loadFeature<W extends Workflow, A extends ActivityInterface>(): Feature<W, A> {
+    return require(path.join(this.absDir, 'feature.ts')).feature;
+  }
+}
+
+export interface RunnerOptions {
+  address: string;
+  namespace: string;
+  taskQueue: string;
+}
+
+export class Runner<W extends Workflow, A extends ActivityInterface> {
+  static async create(source: FeatureSource, options: RunnerOptions) {
+    // Load the feature
+    const feature = source.loadFeature();
+
+    // Connect to client
+    const conn = new Connection({
+      address: options.address,
+    });
+    const client = new WorkflowClient(conn.service, {
+      namespace: options.namespace,
+    });
+
+    // Create and start the worker
+    const workflowsPath =
+      feature.options.workflowsPath ?? require.resolve(path.join(source.absDir, 'feature.workflow.ts'));
+    const worker = await Worker.create({
+      workflowsPath,
+      activities: feature.activities,
+      taskQueue: options.taskQueue,
+    });
+    worker
+      .run()
+      .catch((err) => console.log('Worker failed: ', err))
```

Comment:
Better not to catch this.
It'd be better left as an unhandled rejection or made available for consumption elsewhere.

Reply:
I personally am not a fan of unhandled rejection paths (it leaves details up to the runtime in an unclear way). But I can. I cannot think of a clear way to make it available elsewhere unless I changed `Runner.run` to be a race between this promise and the actual run. Maybe I will do that.

Reply:
Not a fan of unhandled rejection either but it's better than not having the process crash if worker fails.

Reply:
Changing. Unfortunately this takes a longer time because I was forced to split things up. So I have to change the harness in this branch, and then test with a feature from another :-(

Reply:
Changed